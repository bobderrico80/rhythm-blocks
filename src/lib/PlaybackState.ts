enum PlaybackState {
  STOPPED = 'STOPPED',
  PLAYING = 'PLAYING',
}

export default PlaybackState;
