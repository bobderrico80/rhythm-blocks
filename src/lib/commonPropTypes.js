import PropTypes from 'prop-types';
import { noteBlockTypes } from './constants';

export const noteBlock = {
  type: PropTypes.oneOf(Object.values(noteBlockTypes)).isRequired,
  id: PropTypes.string.isRequired,
  svg: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  css: PropTypes.string.isRequired,
  duration: PropTypes.number.isRequired,
};
