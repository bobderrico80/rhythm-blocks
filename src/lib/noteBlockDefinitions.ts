import whole from '../assets/svg/whole.svg';
import halfNote from '../assets/svg/half-note.svg';
import quarterNote from '../assets/svg/quarter-note.svg';
import quaver from '../assets/svg/quaver.svg';
import quarterNoteRest from '../assets/svg/quarter-note-rest.svg';

export enum NoteBlockType {
  WHOLE_NOTE = 'WHOLE_NOTE',
  HALF_NOTE = 'HALF_NOTE',
  QUARTER_NOTE = 'QUARTER_NOTE',
  DOUBLE_8TH_NOTES = 'DOUBLE_8TH_NOTES',
  QUARTER_NOTE_REST = 'QUARTER_NOTE_REST',
}

export interface NoteBlockPattern {
  note: string;
  isRest: boolean;
}

export interface NoteBlockDefinition {
  id?: string;
  type: NoteBlockType;
  svg: string;
  alt: string;
  css: string;
  duration: number;
  pattern: NoteBlockPattern[];
}

export interface NoteBlockDefinitionMap {
  [noteBlockType: string]: NoteBlockDefinition;
}

const NOTE_BLOCK_PATTERNS = {
  WHOLE_NOTE: { note: '1m', isRest: false },
  HALF_NOTE: { note: '2n', isRest: false },
  QUARTER_NOTE: { note: '4n', isRest: false },
  EIGHTH_NOTE: { note: '8n', isRest: false },
  QUARTER_REST: { note: '4n', isRest: true },
};

export const noteBlockDefinitions: NoteBlockDefinition[] = [
  {
    type: NoteBlockType.WHOLE_NOTE,
    svg: whole,
    alt: 'Whole note',
    css: 'whole-note',
    duration: 4,
    pattern: [NOTE_BLOCK_PATTERNS.WHOLE_NOTE],
  },
  {
    type: NoteBlockType.HALF_NOTE,
    svg: halfNote,
    alt: 'Half note',
    css: 'half-note',
    duration: 2,
    pattern: [NOTE_BLOCK_PATTERNS.HALF_NOTE],
  },
  {
    type: NoteBlockType.QUARTER_NOTE,
    svg: quarterNote,
    alt: 'Quarter note',
    css: 'quarter-note',
    duration: 1,
    pattern: [NOTE_BLOCK_PATTERNS.QUARTER_NOTE],
  },
  {
    type: NoteBlockType.DOUBLE_8TH_NOTES,
    svg: quaver,
    alt: 'Two beamed 8th notes',
    css: 'quaver',
    duration: 1,
    pattern: [NOTE_BLOCK_PATTERNS.EIGHTH_NOTE, NOTE_BLOCK_PATTERNS.EIGHTH_NOTE],
  },
  {
    type: NoteBlockType.QUARTER_NOTE_REST,
    svg: quarterNoteRest,
    alt: 'Quarter rest',
    css: 'quarter-note-rest',
    duration: 1,
    pattern: [NOTE_BLOCK_PATTERNS.QUARTER_REST],
  },
];

export const noteBlockDefinitionMap: NoteBlockDefinitionMap = noteBlockDefinitions.reduce(
  (previousMap: NoteBlockDefinitionMap, noteBlockDefinition: NoteBlockDefinition) => ({
    ...previousMap,
    [noteBlockDefinition.type]: noteBlockDefinition,
  }),
  {},
);

export default noteBlockDefinitions;
