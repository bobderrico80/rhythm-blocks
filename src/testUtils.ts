import { ReactElement } from 'react';
import renderer from 'react-test-renderer';

export const createSnapshot = (component: ReactElement<any>) => {
  const snapshot = renderer.create(component);
  return snapshot.toJSON();
};
