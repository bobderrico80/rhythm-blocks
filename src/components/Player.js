import React from 'react';
import PropTypes from 'prop-types';
import { playbackStates } from '../lib/constants';
import { noteBlock } from '../lib/commonPropTypes';
import PlaybackHandler from '../lib/PlaybackHandler';
import styles from './Player.module.css';
import play from '../assets/svg/play.svg';
import stop from '../assets/svg/stop.svg';

const propTypes = {
  measures: PropTypes.arrayOf(
    PropTypes.shape({
      noteBlocks: PropTypes.arrayOf(PropTypes.shape(noteBlock)),
    }),
  ),
  playbackState: PropTypes.oneOf(Object.values(playbackStates)).isRequired,
  onPlaybackStateChange: PropTypes.func.isRequired,
};

const defaultProps = {
  measures: [],
};

class Player extends React.Component {
  playbackHandler = new PlaybackHandler();

  onPlayClick = () => {
    if (!this.props.measures.length || !this.props.measures[0].noteBlocks.length) {
      return;
    }

    if (this.props.playbackState === playbackStates.STOPPED) {
      this.playbackHandler.startPlayback();
    } else {
      this.playbackHandler.stopPlayback();
    }
  };

  scheduleMeasures() {
    const totalMeasureDuration = this.props.measures.reduce((totalDuration, measure) => {
      return (totalDuration += measure.totalDuration);
    }, 0);

    if (totalMeasureDuration > 0) {
      this.playbackHandler.scheduleMeasures(this.props.measures);
    }
  }

  componentDidMount() {
    this.playbackHandler.Transport.on('start', () => {
      this.props.onPlaybackStateChange(playbackStates.PLAYING);
    });

    this.playbackHandler.Transport.on('stop', () => {
      this.props.onPlaybackStateChange(playbackStates.STOPPED);
    });

    this.scheduleMeasures();
  }

  componentDidUpdate() {
    this.scheduleMeasures();
  }

  componentWillUnmount() {
    this.playbackHandler.Transport.off('start');
    this.playbackHandler.Transport.off('stop');
  }

  renderButtonContents() {
    if (this.props.playbackState === playbackStates.PLAYING) {
      return (
        <React.Fragment>
          <img src={stop} alt="Stop" className={styles.stopIcon} />
          <span>Stop</span>
        </React.Fragment>
      );
    }
    return (
      <React.Fragment>
        <img src={play} alt="Play" className={styles.playIcon} />
        <span>Play</span>
      </React.Fragment>
    );
  }

  render() {
    return (
      <button onClick={this.onPlayClick} className={styles.playButton}>
        {this.renderButtonContents()}
      </button>
    );
  }
}

Player.propTypes = propTypes;
Player.defaultProps = defaultProps;

export default Player;
