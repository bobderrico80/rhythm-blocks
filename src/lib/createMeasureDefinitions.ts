import { NoteBlockDefinition } from './noteBlockDefinitions';

export interface MeasureDefinition {
  totalDuration: number;
  noteBlocks: NoteBlockDefinition[];
  beatsPerMeasure: number;
}

const createMeasureDefinitions = (...beatsPerMeasures: number[]): MeasureDefinition[] => {
  return beatsPerMeasures.map(beatsPerMeasure => ({
    totalDuration: 0,
    noteBlocks: [],
    beatsPerMeasure,
  }));
};

export default createMeasureDefinitions;
