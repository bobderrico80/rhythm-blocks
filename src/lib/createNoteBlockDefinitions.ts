import { noteBlockDefinitionMap, NoteBlockDefinition, NoteBlockType } from './noteBlockDefinitions';
import uuid from 'uuid/v4';

const lookupNoteBlockDefinition = (noteBlockType: NoteBlockType): NoteBlockDefinition => {
  return noteBlockDefinitionMap[noteBlockType];
};

const createNoteBlockDefinitions = (
  noteBlockTypes: NoteBlockType | NoteBlockType[],
  noteBlockIdFunction: (noteBlock: NoteBlockDefinition) => any = () => uuid(),
): NoteBlockDefinition[] => {
  return Array.prototype.concat(noteBlockTypes).map(type => {
    const noteBlock = lookupNoteBlockDefinition(type);
    const id = noteBlockIdFunction(noteBlock);
    return { id, ...noteBlock };
  });
};

export default createNoteBlockDefinitions;
