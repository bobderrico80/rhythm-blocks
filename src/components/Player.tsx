import React from 'react';
import PropTypes from 'prop-types';
import PlaybackHandler, { PlaybackState } from '../lib/PlaybackHandler';
import styles from './Player.module.css';
import play from '../assets/svg/player/play.svg';
import stop from '../assets/svg/player/stop.svg';
import { MeasureDefinition } from '../lib/createMeasureDefinitions';

interface PlayerProps {
  measures: MeasureDefinition[];
  playbackState: PlaybackState;
  onPlaybackStateChange: (playbackState: PlaybackState) => void;
}

class Player extends React.Component<PlayerProps> {
  static defaultProps = {
    measures: [],
  };

  playbackHandler = new PlaybackHandler();

  onPlayClick = () => {
    if (!this.props.measures.length || !this.props.measures[0].noteBlocks.length) {
      return;
    }

    if (this.props.playbackState === PlaybackState.STOPPED) {
      this.playbackHandler.startPlayback();
    } else {
      this.playbackHandler.stopPlayback();
    }
  };

  scheduleMeasures() {
    const totalMeasureDuration = this.props.measures.reduce((totalDuration, measure) => {
      return (totalDuration += measure.totalDuration);
    }, 0);

    if (totalMeasureDuration > 0) {
      this.playbackHandler.scheduleMeasures(this.props.measures);
    }
  }

  componentDidMount() {
    this.playbackHandler.Transport.on('start', () => {
      this.props.onPlaybackStateChange(PlaybackState.PLAYING);
    });

    this.playbackHandler.Transport.on('stop', () => {
      this.props.onPlaybackStateChange(PlaybackState.STOPPED);
    });

    this.scheduleMeasures();
  }

  componentDidUpdate() {
    this.scheduleMeasures();
  }

  componentWillUnmount() {
    this.playbackHandler.Transport.off('start');
    this.playbackHandler.Transport.off('stop');
  }

  renderButtonContents() {
    if (this.props.playbackState === PlaybackState.PLAYING) {
      return (
        <React.Fragment>
          <img src={stop} alt="Stop" className={styles.stopIcon} />
          <span>Stop</span>
        </React.Fragment>
      );
    }
    return (
      <React.Fragment>
        <img src={play} alt="Play" className={styles.playIcon} />
        <span>Play</span>
      </React.Fragment>
    );
  }

  render() {
    return (
      <button onClick={this.onPlayClick} className={styles.playButton}>
        {this.renderButtonContents()}
      </button>
    );
  }
}

export default Player;
