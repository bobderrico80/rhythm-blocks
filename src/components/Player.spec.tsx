import React, { ReactElement } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { createSnapshot } from '../testUtils';
import PlaybackState from '../lib/PlaybackState';
import Player, { PlayerProps } from './Player';
import { MeasureDefinition } from '../lib/createMeasureDefinitions';
import createNoteBlockDefinitions from '../lib/createNoteBlockDefinitions';
import { NoteBlockType } from '../lib/noteBlockDefinitions';

jest.mock('../lib/PlaybackHandler');

describe('The <Player /> component', () => {
  const testMeasures: MeasureDefinition[] = [
    {
      totalDuration: 4,
      noteBlocks: createNoteBlockDefinitions([
        NoteBlockType.QUARTER_NOTE,
        NoteBlockType.QUARTER_NOTE,
        NoteBlockType.QUARTER_NOTE,
        NoteBlockType.QUARTER_NOTE,
      ]),
      beatsPerMeasure: 4,
    },
  ];

  let component: ReactElement<any>;
  let wrapper: ShallowWrapper<Readonly<PlayerProps>>;
  let instance: Player;
  let onPlaybackStateChange: jest.Mock;

  beforeEach(() => {
    onPlaybackStateChange = jest.fn();
    component = (
      <Player
        measures={testMeasures}
        playbackState={PlaybackState.STOPPED}
        onPlaybackStateChange={onPlaybackStateChange}
      />
    );
    wrapper = shallow<Player>(component);
    instance = wrapper.instance() as Player;
  });

  describe('with default behavior (in STOPPED state)', () => {
    it('renders as expected', () => expect(createSnapshot(component)).toMatchSnapshot());

    it('calls playbackHandler.scheduleMeasures() with the measures prop on mount', () => {
      expect(instance.playbackHandler.scheduleMeasures).toHaveBeenCalledWith(testMeasures);
    });

    it('sets up Transport `start` event listener on mount', () => {
      expect(instance.playbackHandler.Transport.on).toHaveBeenCalledWith(
        'start',
        expect.any(Function),
      );
    });

    it('sets up Transport `stop` event listener on mount', () => {
      expect(instance.playbackHandler.Transport.on).toHaveBeenCalledWith(
        'stop',
        expect.any(Function),
      );
    });

    it('re-schedules measures when props are updated', () => {
      const updatedMeasures = [
        {
          totalDuration: 4,
          beatsPerMeasure: 4,
          noteBlocks: createNoteBlockDefinitions(NoteBlockType.WHOLE_NOTE),
        },
      ];
      wrapper.setProps({ measures: updatedMeasures });
      expect(instance.playbackHandler.scheduleMeasures).toHaveBeenLastCalledWith(updatedMeasures);
    });

    it('calls onPlaybackStateChange with PLAYING state when `start` event occurs', () => {
      const onMockFn = instance.playbackHandler.Transport.on as jest.Mock;
      onMockFn.mock.calls[0][1](); // Call the 'start' Transport event callback
      expect(onPlaybackStateChange).toHaveBeenCalledWith(PlaybackState.PLAYING);
    });

    it('calls onPlaybackStateChange with STOPPED state when `stop` event occurs', () => {
      const onMockFn = instance.playbackHandler.Transport.on as jest.Mock;
      onMockFn.mock.calls[1][1](); // Call the 'stop' Transport event callback
      expect(onPlaybackStateChange).toHaveBeenCalledWith(PlaybackState.STOPPED);
    });

    describe('button clicking behavior', () => {
      beforeEach(() => {
        wrapper.find('button').simulate('click');
      });

      it('calls playbackHandler.startPlayback()', () => {
        expect(instance.playbackHandler.startPlayback).toHaveBeenCalledTimes(1);
      });

      it('does not call playbackHandler.stopPlayback()', () => {
        expect(instance.playbackHandler.stopPlayback).not.toHaveBeenCalled();
      });
    });

    describe('when unmounting', () => {
      beforeEach(() => {
        wrapper.unmount();
      });

      it('removes `start` event handler', () => {
        expect(instance.playbackHandler.Transport.off).toHaveBeenCalledWith('start');
      });

      it('removes `stop` event handler', () => {
        expect(instance.playbackHandler.Transport.off).toHaveBeenCalledWith('stop');
      });
    });
  });

  describe('during playback', () => {
    beforeEach(() => {
      component = (
        <Player
          measures={testMeasures}
          playbackState={PlaybackState.PLAYING}
          onPlaybackStateChange={onPlaybackStateChange}
        />
      );
      wrapper = shallow(component);
      instance = wrapper.instance() as Player;
    });

    it('renders as expected', () => expect(createSnapshot(component)).toMatchSnapshot());

    describe('button clicking behavior', () => {
      beforeEach(() => {
        wrapper.find('button').simulate('click');
      });

      it('calls playbackHandler.stopPlayback()', () => {
        expect(instance.playbackHandler.stopPlayback).toHaveBeenCalledTimes(1);
      });

      it('does not call playbackHandler.startPlayback()', () => {
        expect(instance.playbackHandler.startPlayback).not.toHaveBeenCalled();
      });
    });
  });

  describe('with no measures', () => {
    beforeEach(() => {
      component = (
        <Player
          measures={[]}
          playbackState={PlaybackState.PLAYING}
          onPlaybackStateChange={onPlaybackStateChange}
        />
      );
      wrapper = shallow(component);
      instance = wrapper.instance() as Player;
    });

    it('does not call playbackHandler.scheduleMeasures()', () => {
      expect(instance.playbackHandler.scheduleMeasures).not.toHaveBeenCalled();
    });

    describe('button clicking behavior', () => {
      beforeEach(() => {
        wrapper.find('button').simulate('click');
      });

      it('does not call playbackHandler.startPlayback()', () => {
        expect(instance.playbackHandler.startPlayback).not.toHaveBeenCalled();
      });

      it('does not call playbackHandler.stopPlayback()', () => {
        expect(instance.playbackHandler.stopPlayback).not.toHaveBeenCalled();
      });
    });
  });

  describe('with empty measures', () => {
    beforeEach(() => {
      component = (
        <Player
          measures={[
            {
              totalDuration: 0,
              noteBlocks: [],
              beatsPerMeasure: 4,
            },
          ]}
          playbackState={PlaybackState.PLAYING}
          onPlaybackStateChange={onPlaybackStateChange}
        />
      );
      wrapper = shallow(component);
      instance = wrapper.instance() as Player;
    });

    it('does not call playbackHandler.scheduleMeasures()', () => {
      expect(instance.playbackHandler.scheduleMeasures).not.toHaveBeenCalled();
    });

    describe('button clicking behavior', () => {
      beforeEach(() => {
        wrapper.find('button').simulate('click');
      });

      it('does not call playbackHandler.startPlayback()', () => {
        expect(instance.playbackHandler.startPlayback).not.toHaveBeenCalled();
      });

      it('does not call playbackHandler.stopPlayback()', () => {
        expect(instance.playbackHandler.stopPlayback).not.toHaveBeenCalled();
      });
    });
  });
});
