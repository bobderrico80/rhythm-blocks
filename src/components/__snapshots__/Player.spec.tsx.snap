// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Player /> component during playback renders as expected 1`] = `
<button
  className="playButton"
  onClick={[Function]}
>
  <img
    alt="Stop"
    className="stopIcon"
    src="stop.svg"
  />
  <span>
    Stop
  </span>
</button>
`;

exports[`The <Player /> component with default behavior (in STOPPED state) renders as expected 1`] = `
<button
  className="playButton"
  onClick={[Function]}
>
  <img
    alt="Play"
    className="playIcon"
    src="play.svg"
  />
  <span>
    Play
  </span>
</button>
`;
